File: python-protobuf/read_message.py
#! /usr/bin/pythin

from build import addressbook_pb2
import sys

# Iterate through all the people in the AddressBook and print info about them
def ListPeople(address_book):
    for person in address_book.people:
        print("Person ID:", person.id)
        print("  Name:", person.name)
        if person.email != "":
            print("  E-mail address:", person.email)

        for phone_number in person.phones:
            if phone_number.type == addressbook_pb2.Person.PhoneType.MOBILE:
                print("  Mobile phone #: ", end="")
            elif phone_number.type == addressbook_pb2.Person.PhoneType.HOME:
                print("  Home phone #: ", end="")
            elif phone_number.type == addressbook_pb2.Person.PhoneType.WORK:
                print("  Work phone #: ", end="")
            print(phone_number.number)


# Read the entire address book from a file and print all the information inside
def main():
    if len(sys.argv) != 2:
        print("Usage:", sys.argv[0], "ADDRESS_BOOK_FILE")
        sys.exit(-1)

    # Create an instance of the protobuf message
    address_book = addressbook_pb2.AddressBook()

    # Read the existing address book
    with open(sys.argv[1], "rb") as f:
        address_book.ParseFromString(f.read())
    
    ListPeople(address_book)


if __name__ == "__main__":
    main()

File: csharp-protobuf/ReadMessage.cs
using System;
using System.IO;

using Google.Protobuf;
using static Protos.Addressbook.Person.Types;

namespace Protos.Addressbook
{
    class ReadMessage
    {
        private static void Print(AddressBook addressBook)
        {
            foreach (Person person in addressBook.People)
            {
                Console.WriteLine("Person ID: {0}", person.Id);
                Console.WriteLine("  Name: {0}", person.Name);
                if (person.Email != "")
                {
                    Console.WriteLine("  E-mail address: {0}", person.Email);
                }

                foreach (PhoneNumber phoneNumber in person.Phones)
                {
                    switch (phoneNumber.Type)
                    {
                        case PhoneType.Mobile:
                            Console.Write("  Mobile phone #: ");
                            break;
                        case PhoneType.Home:
                            Console.Write("  Home phone #: ");
                            break;
                        case PhoneType.Work:
                            Console.Write("  Work phone #: ");
                            break;
                    }
                    Console.WriteLine(phoneNumber.Number);
                }
            }
        }

        public static int Main(string[] args)
        {
            if (args.Length != 1)
            {
                Console.Error.WriteLine("Usage: ReadMessage ADDRESS_BOOK_FILE");
                return 1;
            }

            if (!File.Exists(args[0]))
            {
                Console.WriteLine("{0} does not exist. Add a person to create the file first.", args[0]);
                return 0;
            }

            // Read the existing address book
            using (Stream stream = File.OpenRead(args[0]))
            {
                AddressBook addressBook = AddressBook.Parser.ParseFrom(stream);
                Print(addressBook);
            }
            return 0;
        }
    }
}
